### ISSUE TITLE
mvel 2.4.3 does not export OSGi package which was exported in 2.4.2

### ISSUE THREAD

**Issue by davsclaus** [mvel/mvel#178]
mvel 2.4.2 works in OSGi where you can import the package `org.mvel2.asm.util` package which some 3rd party projects like jBPMN and OptaPlanner uses. So we in Apache Camel cannot upgrade to mvel 2.4.3 from 2.4.2.

Here is an output from out validator

[WARNING] Feature resolution failed for [camel-jbpm/3.0.0.SNAPSHOT]
Message: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=camel-jbpm; type=karaf.feature; version=3.0.0.SNAPSHOT; filter:="(&(osgi.identity=camel-jbpm)(type=karaf.feature)(version>=3.0.0.SNAPSHOT))" [caused by: Unable to resolve camel-jbpm/3.0.0.SNAPSHOT: missing requirement [camel-jbpm/3.0.0.SNAPSHOT] osgi.identity; osgi.identity=org.apache.camel.camel-jbpm; type=osgi.bundle; version="[3.0.0.SNAPSHOT,3.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve org.apache.camel.camel-jbpm/3.0.0.SNAPSHOT: missing requirement [org.apache.camel.camel-jbpm/3.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.kie.server.api.model.instance)(version>=7.14.0)(!(version>=8.0.0)))" [caused by: Unable to resolve org.kie.server.api/7.14.0.201811070853: missing requirement [org.kie.server.api/7.14.0.201811070853] osgi.wiring.package; filter:="(osgi.wiring.package=org.drools.core.command)" [caused by: Unable to resolve org.drools.core/7.14.0.201811070642: missing requirement [org.drools.core/7.14.0.201811070642] osgi.wiring.package; filter:="(osgi.wiring.package=org.mvel2.asm.util)"]]]]
Repositories: {

**Comment by mariofusco** [comment]
Claus,

I'm honestly not understanding where the problem with that missing package comes from. mvel is not OSGfied itself and I cannot find any reference to the package 'org.mvel2.asm.util' in drools, jbpm or optaplanner code base. Can you please clairfy?

Thanks,
Mario

**Comment by davsclaus** [comment]
Here is a list of exported packages in mvel 2.4.2 vs 2.4.3
```
karaf@root()> exports -b 46
Package Name                              │ Version     │ ID │ Bundle Name
──────────────────────────────────────────┼─────────────┼────┼────────────
org.mvel2.asm.commons                     │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.asm.signature                   │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.asm.tree.analysis               │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.asm.tree                        │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.asm.util                        │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.asm.xml                         │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.asm                             │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.ast                             │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.compiler                        │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.conversion                      │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.debug                           │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.integration.impl                │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.integration                     │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.jsr223                          │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.math                            │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.optimizers.dynamic              │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.optimizers.impl.asm             │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.optimizers.impl.refl.collection │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.optimizers.impl.refl.nodes      │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.optimizers.impl.refl            │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.optimizers                      │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.sh.command.basic                │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.sh.command.file                 │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.sh.text                         │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.sh                              │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.templates.res                   │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.templates.util.io               │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.templates.util                  │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.templates                       │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2.util                            │ 2.4.2.Final │ 46 │ org.mvel2
org.mvel2                                 │ 2.4.2.Final │ 46 │ org.mvel2
karaf@root()> exports -b 45
Package Name                              │ Version     │ ID │ Bundle Name
──────────────────────────────────────────┼─────────────┼────┼────────────
org.mvel2.asm.signature                   │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.asm                             │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.ast                             │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.compiler                        │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.conversion                      │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.debug                           │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.integration.impl                │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.integration                     │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.jsr223                          │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.math                            │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.optimizers.dynamic              │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.optimizers.impl.asm             │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.optimizers.impl.refl.collection │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.optimizers.impl.refl.nodes      │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.optimizers.impl.refl            │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.optimizers                      │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.sh.command.basic                │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.sh.command.file                 │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.sh.text                         │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.sh                              │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.templates.res                   │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.templates.util.io               │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.templates.util                  │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.templates                       │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2.util                            │ 2.4.3.Final │ 45 │ org.mvel2
org.mvel2                                 │ 2.4.3.Final │ 45 │ org.mvel2
```

As you can see 2.4.2 exports more of the asm packages and 2.4.3 does not

**Comment by davsclaus** [comment]
Ah browsing your source code in github and looking at the tags for 2.4.2 and 2.4.3 and 2.4.4 then I can see that you did many code changes in the asm package and all those sub packages are gone. So there has been some bigger code changes going on. Maybe that should have gone into a 2.5.0 release instead?

**Comment by davsclaus** [comment]
Ah okay looks like you upgraded from asm6 to asm7 in that "patch" release

**Comment by mariofusco** [comment]
I upgraded the embedded asm indeed, and this caused (among other things) the elimination of that org.mvel2.asm.util package, but still I don't understand why this broke OSGi. Can you please clarify and explain what I should do to fix it?

**Comment by davsclaus** [comment]
Yeah OSGi can be tricky, looks like its drools-core 7.14.0.Final that has an OSGi import on that mvel asm package, which it should really not do. Also with OSGi you can have mvn dependency tree vs OSGi imports being different.

**Comment by davsclaus** [comment]
Looks like drools-core 7.17.0.Final is better.

**Comment by davsclaus** [comment]
And also optaplanner 7.14.0 had a problem which they seems to have fixed in 7.17.0

**Comment by davsclaus** [comment]
So this was more of a problem with 3rd party projects that mistakenly import asm packages from mvel. And mvel should not really export them, as you use them for internal purposes. Only export the packages that clients of mvel need to use.
